### DISCUSSION TITLE
no GUI on plugins on Linux

### DISCUSSION THREAD

**Discussion by evertvorster** [audacity/audacity#3476] [Development]
### Bug description

Hi there!
When building against system wxWidgets, no GUI is displayed on Calf plugins.
I suspect that all plugins are affected, but have not tested this theory.

These plugins display their UI when I use the wxWidgets that conan builds when "local" is specified. This library links against Gtk2, and there does not seem to be a way to change that.

I have now tried several iterations of wxWidgets, both the 3.2 version that Arch Linux ships with now, and versions of wxWidgets back as far as 3.1.4

All of them show the following error in the console:
suil error: Unable to wrap UI type <http://lv2plug.in/ns/extensions/ui#GtkUI> as type <http://lv2plug.in/ns/extensions/ui#Gtk3UI>

This error only shows up when Audacity is built with "Debug" target. I am unable to get more information on this error message. Is there a way to increase the verbosity of the debug messages?

### Steps to reproduce

1. Open a plugin effect that uses a GUI.

### Expected behavior

The GUI of the effect to appear

### Actual behavior

A box full of sliders. 

### Audacity Version

current master / alpha / nightly build

### Operating system

Linux

### Additional context

I am quite happy to do testing and build from sources to test any fixes. ;)

**Comment by crsib** [comment]
I wonder if this issue is related to https://github.com/calf-studio-gear/calf/issues/248

Probably something in our configuration does not allow SUIL to host GTK2 in GTK3

**Comment by evertvorster** [comment]
"Probably something in our configuration does not allow SUIL to host GTK2 in GTK3"
That would be weird, as the point of suil is to allow one toolkit to be displayed in another. What's the easiest way to check?

Thanks for the pointer, it pointed me in something that I have not tested yet. Currently all the system libraries tested were gtk3, and the supplied one was gtk2.

So, now I am building a gtk2 version 3.2.0 in the system, and will build audacity against that, to prove or disprove whether the problem is gtk2/3, or whether the library is local or system. 

If I remember correctly, I was able to get the GUI on gtk3 some time ago. You mentioned that you are moving away from the wxwidgets toolkit. Which toolkits are being considered for replacement?

**Comment by evertvorster** [comment]
Forgive my ignorance... The moving away from wxwidgets report was in calf, not audacity. Are there any plans to move Audacity from the wxwidgets toolkit?

**Comment by crsib** [comment]
We have plans to migrate as well, this will take quite a time though

**Comment by evertvorster** [comment]
Hmm. 
I can build Audacity against system gtk2 (version 3.2.0), and then the calf plugins work, after a fashion. Some of the visual effects are gone, and there is no play button. I believe I saw a bug report for the missing UI effects here. 

The creator of the suil library said the following on https://github.com/lv2/lv2/wiki#avoid-gui-frameworks:
"gtk2 + gtk3 cannot exist in the same memory-space, likewise Qt4 + Qt5. This means a gtk2 host cannot load plugins with a gtk3 GUI, nor can you load a Qt4 plugin at the same time as plugin using a Qt5 GUI."

This would explain why suil won't display gtk2 effects when Audacity is built with gtk3, no matter which version. 
Having Audacity use gtk2 actually avoids some other issues, like the text not fitting in buttons. The next thing for me to test would be to see if the latest released branch of Audacity also has the missing UI elements when compiled against system gtk2.

**Comment by evertvorster** [comment]
On the Qt branch of Audacity, would it be possible to get Audacity to link against the QT version of wxWidgets?
Since the LV2 plugins are built against GTK, this might enable them to work for a little while longer. 

In Arch it's called wxwidgets-qt5, and it installs the following files:
```
wxwidgets-qt5 /usr/bin/wx-config-qt
wxwidgets-qt5 /usr/lib/
wxwidgets-qt5 /usr/lib/cmake/
wxwidgets-qt5 /usr/lib/cmake/wxWidgetsQt/
wxwidgets-qt5 /usr/lib/cmake/wxWidgetsQt/wxWidgetsQtConfig.cmake
wxwidgets-qt5 /usr/lib/cmake/wxWidgetsQt/wxWidgetsQtConfigVersion.cmake
wxwidgets-qt5 /usr/lib/cmake/wxWidgetsQt/wxWidgetsQtTargets-none.cmake
wxwidgets-qt5 /usr/lib/cmake/wxWidgetsQt/wxWidgetsQtTargets.cmake
wxwidgets-qt5 /usr/lib/libwx_qtu_adv-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_adv-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_adv-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_aui-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_aui-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_aui-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_core-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_core-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_core-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_gl-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_gl-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_gl-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_html-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_html-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_html-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_media-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_media-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_media-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_propgrid-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_propgrid-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_propgrid-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_qa-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_qa-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_qa-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_ribbon-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_ribbon-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_ribbon-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_richtext-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_richtext-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_richtext-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_stc-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_stc-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_stc-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/libwx_qtu_xrc-3.2.so
wxwidgets-qt5 /usr/lib/libwx_qtu_xrc-3.2.so.0
wxwidgets-qt5 /usr/lib/libwx_qtu_xrc-3.2.so.0.2.1
wxwidgets-qt5 /usr/lib/wx/
wxwidgets-qt5 /usr/lib/wx/config/
wxwidgets-qt5 /usr/lib/wx/config/qt-unicode-3.2
wxwidgets-qt5 /usr/lib/wx/include/
wxwidgets-qt5 /usr/lib/wx/include/qt-unicode-3.2/
wxwidgets-qt5 /usr/lib/wx/include/qt-unicode-3.2/wx/
wxwidgets-qt5 /usr/lib/wx/include/qt-unicode-3.2/wx/setup.h
wxwidgets-qt5 /usr/share/
wxwidgets-qt5 /usr/share/licenses/
wxwidgets-qt5 /usr/share/licenses/wxwidgets-qt5/
wxwidgets-qt5 /usr/share/licenses/wxwidgets-qt5/LICENSE
```

**Comment by crsib** [comment]
The plan is to migrate to Qt :-)

**Reply by evertvorster** [reply]
So, for the qt branch of audacity we will eventually not use wxWidgets at all?

**Reply by crsib** [reply]
That's the plan ☺️

**Reply by evertvorster** [reply]
It goes without saying that I can't wait! Thanks for all the effort you are putting in.
